man/man3/orbit.3
.TH ORBIT 3NEMO "24 May 2025" 

.SH "NAME"
read_orbit, write_orbit, allocate_orbit, list_orbit \- basic orbit i/o routines
.SH SYNOPSIS
.nf
.B #include <stdinc.h>
.B #include <potential.h>
.B #include <orbit.h>
.PP
.B allocate_orbit(optr,ndim,nsteps)
.B orbitptr *optr;
.B int ndim;
.B int nsteps;
.PP
.B write_orbit(outstr,optr)
.B stream outstr;
.B orbitptr optr;
.PP
.B read_orbit(instr,optr)
.B stream instr;
.B orbitptr *optr;
.PP
.B list_orbit(optr,tstart,tend,n)
.B orbitptr optr;
.B double tstart;
.B double tend;
.B int n;
.fi

.SH "DESCRIPTION"
\fIorbit(3NEMO)\fP provide a few utility functions which handle the i/o of
\fIorbit(5NEMO)\fP datafile. They subsequently call lower level filestructure 
routines, as described in \fIfilestruct(3NEMO)\fP.
.PP
\fIwrite_orbit\fP writes an orbit to an output stream, using standard
binary filestructure.
.PP
\fIread_orbit\fP reads an orbit from an input stream into a pointer
to an orbit. If that pointer is NULL, i.e. has not been initialized,
this routine will take care of the allocation too.
.PP
\fIallocate_orbit\fP allocates an orbit. Note that \fIoptr\fP is
a pointer to an orbit, on entry normally initialized to NULL
to signal that a new orbit must be allocated. If  \fIoptr\fP was already
assigned a value (previous allocation), \fIallocate_orbit\fP checks if
this one contains enough space. If not, new space is claimed and old
space may be freed. Normally application programs do not need
to call this routine, since \fIread_orbit\fP takes care of the allocation.
.PP
\fIlist_orbit\fP lists the coordinates of an orbit between two given
times.

.SH "SEE ALSO"
\fIorbit(5NEMO)\fP

.SH "BUGS"
No proper dynamic allocation of orbits yet; space has to be allocated
before it's used in the orbit calculation.

.SH "AUTHOR"
Peter Teuben

.SH "FILES"
.nf
.ta +2.5i
~/src/pjt/orbit  	authors original sources
.fi

.SH "HISTORY"
.nf
.ta +1.25i +4.5i
13-Jul-87	V1.0: Original created	PJT
28-jul-87	V2.0: new orbit(5) structure	PJT
24-may-92	V3.0: added potential to orbit	PJT
dec-2019	V4.0: support for storing Phi/Acc	PJT
jul-2023	V5.1: support storing key	PJT
.fi


src/orbit/io/orbit.c
/* orbit.c - write_orbit, read_orbit, allocate_orbit, copy_orbit, list_orbit */

/*------------------------------------------------------------------------------
 * ORBIT      proposed format for orbit paths
 *              documentation in orbit.3
 *
 *  13-Jul-87   V1.0 created                            P.J. Teuben
 *  28-jul-87   V2.0 new orbit(5)                               PJT
 *   3-May-88   ---- added few dprintf()'s and data history     PJT
 *   2-jun-88   ---- newly compiled for new filestruct          PJT
 *                      get/put_hist now
 *  22-may-90   new copy_orbit, debug read_orbit                PJT
 *  11-jun-90   list_orbit output
 *  19-dec-90   read_orbit uses allocate_orbit for allocate()   PJT
 *		  but now there is no method to save allocated space PJT
 *  24-may-92   V3.0 added potential                            PJT
 *  23-nov-93   V3.1 but fixed rather emberassing bug           pjt
 *   6-dec-93	V3.2 added maxsteps				pjt
 *  22-feb-95   V3.2a ansi					pjt
 *  17-apr-95   V3.2b compacted needed header files		pjt
 *  19-apr-96       c initialize strings to "" instead of NULL  pjt
 *   1-mar-03   V3.3  added stuff for iom_err			pjt
 *  25-jul-13   V4.0  added support for key                     pjt
 *  12-nov-2015 V5.0  some support for split Pos/Vel            pjt
 *  10-dec-2019 V5.1  optional support for PHI/ACC              pjt
 *   5-oct-2021 V5.2  fix for 3D orbits and printing PHI/ACC    pjt
 *   8-feb-2022       free the extra PHI/ACC                    pjt
 *  10-dec-2023       report etot in list_orbit()               pjt
 *------------------------------------------------------------------------------
 */

#include <stdinc.h>
#include <orbit.h>

#define ISSTRING(s) ( (s)!=NULL && *(s)!=0 )

/*------------------------------------------------------------------------------
 *  WRITE_ORBIT: writes out one orbit
 *------------------------------------------------------------------------------
 */

void write_orbit (stream outstr, orbitptr optr)
{
    /* No history written here, user should do this */
    put_set (outstr,OrbitTag);
        put_set (outstr,ParametersTag);
            put_data (outstr,NdimTag,    IntType, &(Ndim(optr)),   0);
            put_data (outstr,MassTag,    RealType,&(Masso(optr)),   0);
            put_data (outstr,KeyTag,     IntType, &(Key(optr)),   0);
            put_data (outstr,IOMTag,     RealType,  IOM(optr), Ndim(optr), 0);
            put_data (outstr,IOMERRTag,  RealType,  IOMERR(optr), Ndim(optr), 0);
	    /* note we don't write MAXsteps(optr) to disk */
            put_data (outstr,NstepsTag,  IntType, &(Nsteps(optr)), 0);
        put_tes (outstr,ParametersTag);

        put_set (outstr,PotentialTag);
            if (ISSTRING(PotName(optr))) put_string (outstr, PotNameTag, PotName(optr));
            if (ISSTRING(PotPars(optr))) put_string (outstr, PotParsTag, PotPars(optr));
            if (ISSTRING(PotFile(optr))) put_string (outstr, PotFileTag, PotFile(optr));
        put_tes (outstr,PotentialTag);

        put_set (outstr,PathTag);
            put_data (outstr,TimePathTag, RealType, 
                                TimePath(optr), Nsteps(optr), 0);
            put_data (outstr,PhasePathTag,RealType, 
                                PhasePath(optr), Nsteps(optr), 2, Ndim(optr), 0);
#ifdef ORBIT_PHI
	    put_data (outstr,PhiPathTag,RealType,
		      PhiPath(optr), Nsteps(optr), 0);
	    put_data (outstr,AccPathTag,RealType,
		      AccPath(optr), Nsteps(optr), Ndim(optr), 0);
	    
#endif	    
        put_tes (outstr,PathTag);
    put_tes (outstr,OrbitTag);
}

/*------------------------------------------------------------------------------
 * READ_ORBIT: try and read another orbit
 *            returns 0 if not, or an error occurred,  1 if OK
 *------------------------------------------------------------------------------
 */
 
int read_orbit (stream instr, orbitptr *optr)
{
    int  maxsize, ndim, nsteps;
    orbitptr otmp;

    get_history(instr);     /* always scan for history; allow sandwiched */

    // @todo if Orbit not found, continue until one if found or EOF
    if (!get_tag_ok (instr,OrbitTag))
        return 0;                      /* not another orbit available */
        
    if (*optr!=NULL)
        maxsize = Size(*optr);
    else
        maxsize = 0;
        
    get_set (instr,OrbitTag);
        get_set (instr,ParametersTag);
            get_data (instr,NdimTag,    IntType, &ndim,   0);
            get_data (instr,NstepsTag,  IntType, &nsteps, 0);
            if (ndim*nsteps > maxsize) {	      /* allocate more space */
                if (*optr) free_orbit(*optr);         /* free the old one    */
                allocate_orbit(&otmp,ndim,nsteps);
                *optr = otmp;
		Size(*optr) = ndim*nsteps;
		dprintf(1,"Reallocated orbit with %d space\n",ndim*nsteps);
            } else {
                Ndim(*optr) = ndim;
                Nsteps(*optr) = nsteps;
            }
            get_data (instr, MassTag, RealType, &(Masso(*optr)), 0);
	    if (get_tag_ok(instr,KeyTag))
	      get_data (instr, KeyTag, IntType, &Key(*optr), 0);
	    else
	      Key(*optr) = 0;
            get_data (instr, IOMTag, RealType, IOM(*optr), Ndim(*optr), 0);
            if (get_tag_ok(instr,IOMERRTag))
                get_data (instr, IOMERRTag, RealType, IOMERR(*optr), Ndim(*optr), 0);
            else
                IE1(*optr) = IE3(*optr) = IE3(*optr) = 0.0;
        get_tes (instr,ParametersTag);

#if 0
        PotName(*optr) = NULL;
        PotPars(*optr) = NULL;
        PotFile(*optr) = NULL;
#else
        PotName(*optr) = "";
        PotPars(*optr) = "";
        PotFile(*optr) = "";
#endif        
	if (get_tag_ok(instr,PotentialTag)) {
            get_set(instr,PotentialTag);
            if(get_tag_ok(instr,PotNameTag)) 
                PotName(*optr) = get_string(instr,PotNameTag);
            if(get_tag_ok(instr,PotParsTag)) 
                PotPars(*optr) = get_string(instr,PotParsTag);
            if(get_tag_ok(instr,PotFileTag)) 
                PotFile(*optr) = get_string(instr,PotFileTag);
            get_tes(instr,PotentialTag);
	} 

        get_set (instr,PathTag);
            get_data (instr,TimePathTag,    RealType, 
                      TimePath(*optr), Nsteps(*optr), 0);
            get_data (instr,PhasePathTag,RealType, 
                      PhasePath(*optr), Nsteps(*optr), 2, Ndim(*optr), 0);
#ifdef ORBIT_PHI
	    dprintf(1,"Reading orbit Phi/Acc\n");
	    if (get_tag_ok(instr, PhiPathTag)) {
	      get_data (instr,PhiPathTag,    RealType, 
                      PhiPath(*optr), Nsteps(*optr), 0);
	      get_data (instr,AccPathTag,RealType, 
		      AccPath(*optr), Nsteps(*optr), Ndim(*optr), 0);
	      dprintf(2,"Found Phi/Acc %g %g\n", PhiPath(*optr)[0], AccPath(*optr)[0]); 
	    }
#endif	    
        get_tes (instr,PathTag);
    get_tes (instr,OrbitTag);
    return 1;
}

void free_orbit(orbitptr optr)
{
    free( (char *) TimePath(optr) );
    free( (char *) PhasePath(optr) );
    free( (char *) IOM(optr) );
    free( (char *) IOMERR(optr) );
#ifdef ORBIT_PHI
    free( (char *) PhiPath(optr) );
    free( (char *) AccPath(optr) );
#endif    
    free( (char *) optr );
}

/*------------------------------------------------------------------------------
 * ALLOCATE_ORBIT: allocate space for an orbit path
 *      returns 0 if failure due to lack of memory or so
 *      returns 1 if all seems OK.
 *------------------------------------------------------------------------------
 */
 
int allocate_orbit(orbitptr *optr, int ndim, int nsteps)
{
    *optr = (orbitptr ) allocate(sizeof(orbit));
    dprintf (1,"Allocate_Orbit @ %x with ndim=%d nsteps=%d\n",
                *optr, ndim, nsteps);
    Ndim(*optr) = ndim;
    IOM(*optr)    = (real *) allocate(ndim*sizeof(real));
    dprintf (2,"   IOM allocated @ %x\n",IOM(*optr)); 
    IOMERR(*optr) = (real *) allocate(ndim*sizeof(real));
    dprintf (2,"   IOMERR allocated @ %x\n",IOMERR(*optr)); 
    Nsteps(*optr) = nsteps;
    TimePath(*optr) = (real *) allocate(nsteps*sizeof(real));
    dprintf (2,"   Timepath allocated @ %x\n",TimePath(*optr));
    PhasePath(*optr) = (real* ) allocate(nsteps*ndim*2*sizeof(real));
    dprintf (2,"  Phasepath allocated @ %x\n",PhasePath(*optr));
#ifdef ORBIT_PHI
    PhiPath(*optr) = (real *) allocate(nsteps*sizeof(real));
    AccPath(*optr) = (real* ) allocate(nsteps*ndim*sizeof(real));
#endif    
    Size(*optr) = ndim*nsteps;
    MAXsteps(*optr) = nsteps;  /* ??? */

    return 1;      /* succes */
}

/*------------------------------------------------------------------------------
 * COPY_ORBIT: copy an orbit, assumes destination has been allocated.
 *------------------------------------------------------------------------------
 */
 
void copy_orbit(orbitptr iptr, orbitptr optr)
{
        int i, nsteps;
        
        if (optr==NULL) error("copy_orbit: Output orbit not allocated");
	
	if (Nsteps(optr) < Nsteps(iptr)) 
	  error("copy_orbit: not enough space in orbit to copy to: out->%d < in->%d",
		Nsteps(optr),Nsteps(iptr));
        
        Ndim(optr) = Ndim(iptr);
        CoordSys(optr) = CoordSys(iptr);
        Masso(optr) = Masso(iptr);
	Key(optr) = Key(iptr);
        IOM(optr) = IOM(iptr);                 /* needs a copy, not a link ?? */
        IOMERR(optr) = IOMERR(iptr);
        nsteps = Nsteps(optr) = Nsteps(iptr);
        for (i=0; i<nsteps; i++) {
            Torb(optr,i) = Torb(iptr,i);
            Xorb(optr,i) = Xorb(iptr,i);
            Yorb(optr,i) = Yorb(iptr,i);
            Zorb(optr,i) = Zorb(iptr,i);
            Uorb(optr,i) = Uorb(iptr,i);
            Vorb(optr,i) = Vorb(iptr,i);
            Worb(optr,i) = Worb(iptr,i);
#ifdef ORBIT_PHI	    
	    Porb(optr,i) = Porb(iptr,i);
	    AXorb(optr,i)= AXorb(iptr,i);
	    AYorb(optr,i)= AYorb(iptr,i);
	    AZorb(optr,i)= AZorb(iptr,i);
#endif
        }
}

/*------------------------------------------------------------------------------
 * LIST_ORBIT:  simply list the orbits positions and velocities
 *------------------------------------------------------------------------------
 */
 
void list_orbit (orbitptr optr, double tstart, double tend, int n, string f)
{
    int i, kount;
    char fmt7[256];
    char fmt12[256];
    real etot = 0.0;

    sprintf(fmt12, "%%d %s %s %s %s %s %s %s  %s %s %s %s %s\n",f,f,f,f,f,f,f, f,f,f,f,f);
    sprintf(fmt7  ,"%%d %s %s %s %s %s %s %s\n",             f,f,f,f,f,f,f);
        
    dprintf (0,"Total number of steps = %d\n",Nsteps(optr));
    dprintf (0,"Mass = %f \n",Masso(optr));
    dprintf (0,"Key = %d \n",Key(optr));
    dprintf (0,"Integrals of Motion (IOM) = ");
    for (i=0; i<Ndim(optr); i++)
        dprintf (0," %f ",*(IOM(optr)+i));
#ifdef PJT312
    dprintf (0," ( ",*(IOM(optr)+i));
#else
    dprintf (0," ( ");
#endif
    for (i=0; i<Ndim(optr); i++)
        dprintf (0," %f ",*(IOMERR(optr)+i));
    dprintf (0," )\n");
#ifdef ORBIT_PHI
    dprintf (0,"ORBIT_PHI enabled %g %g\n",PhiPath(optr)[0], Porb(optr,0));
#else
    dprintf (0,"ORBIT_PHI disabled\n");    
#endif
    dprintf (0,"Potential: Name: %s Pars: %s File: %s\n",
        PotName(optr), PotPars(optr), PotFile(optr));
    kount = 0;
    for (i=0; i<Nsteps(optr); i++) {
        if ((tstart<Torb(optr,i)) && (Torb(optr,i)<tend)) {
#ifdef ORBIT_PHI	  
            etot = 0.5*(sqr(Uorb(optr,i))+sqr(Vorb(optr,i))+sqr(Worb(optr,i))) + Porb(optr,i);
#endif	  
            if (kount++ == 0)
#ifdef ORBIT_PHI
                printf (fmt12,
			i,Torb(optr,i),Xorb(optr,i),Yorb(optr,i),Zorb(optr,i),
			Uorb(optr,i),Vorb(optr,i),Worb(optr,i),
			Porb(optr,i),
			AXorb(optr,i),AYorb(optr,i),AZorb(optr,i),
			etot);
#else	      
                printf (fmt7,
			i,Torb(optr,i),Xorb(optr,i),Yorb(optr,i),Zorb(optr,i),
			Uorb(optr,i),Vorb(optr,i),Worb(optr,i));
#endif	    
            if (kount==n)
                kount=0;
        }
    }
}


#ifdef TESTBED
#include <getparam.h>

string defv[] = {
        "name=???\n    Filename",
        "mode=w\n      R or W",
        "VERSION=1.3\n 10-dec-2019 pjt",
        NULL,
};

string usage = "orbit testbed";

void ini_orbit(orbitptr);

nemo_main()
{
    orbitptr    optr=NULL;
    stream      instr,outstr;
    string      fname=getparam("name");
    
    if (strcmp(getparam("mode"),"w")==0) {              /* write test */
        printf ("WRITING test\n");
        outstr = stropen (fname,"w");
        allocate_orbit(&optr,3,4);      /* 3-D  4 steps */
        ini_orbit (optr);
        write_orbit (outstr,optr);
        strclose(outstr);
    } else {
        printf ("READING test\n");
        instr = stropen (fname,"r");
        while (read_orbit(instr,&optr)) {
                list_orbit(optr,-10.0,10.0,1,"%g");
        }
        strclose(instr);
    }
}

void ini_orbit (optr)
orbitptr optr;
{
    int i;
        
    Masso(optr)  = 1.0;
    Key(optr) = 1;
    I1(optr) = -1.0;
    I2(optr) = 0.0;
    I3(optr) = 0.0;

    PotName(optr) = "";
    PotPars(optr) = "";
    PotFile(optr) = "";

    for (i=0; i<Nsteps(optr); i++) {
        Torb(optr,i) = 0.1*i;
        Xorb(optr,i) = 0.1*i+1.0;
        Yorb(optr,i) = 0.1*i+2.0;
        Zorb(optr,i) = 0.1*i+3.0;
        Uorb(optr,i) = -Xorb(optr,i);
        Vorb(optr,i) = -Yorb(optr,i);
        Worb(optr,i) = -Zorb(optr,i);
    }
}
#endif
